packages/datadog-instrumentations/src/mongodb-core.js

```diff
@@ -179,3 +189,28 @@ function instrument (operation, command, ctx, args, server, ns, ops, options = {
     }
   })
 }
+
+function instrumentPromise (operation, command, ctx, args, server, ns, ops, options = {}) {
+  const name = options.name || (ops && Object.keys(ops)[0])
+
+  const serverInfo = server && server.s && server.s.options
+  const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+  return asyncResource.runInAsyncScope(() => {
+    startCh.publish({ ns, ops, options: serverInfo, name })
+
+    const promise = command.apply(ctx, args)
+
+    promise.then(function (res) {
+      finishCh.publish()
+      return res
+    }, function (err) {
+      errorCh.publish(err)
+      finishCh.publish()
+
+      return Promise.reject(err)
+    })
+
+    return promise
```

Comment:
Is it intentional that you're returning `promise` here instead of the result of `promise.then`? That'd mean that awaits/thens on the return value of the function might execute before channel publications executed.

Reply:
Intentional yes, but tbh just based in other use cases that I found in the codebase, like [cucumber](https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-instrumentations/src/cucumber.js#L183) or [cassandra-driver](https://github.com/DataDog/dd-trace-js/blob/master/packages/datadog-instrumentations/src/cassandra-driver.js#L70) instrumentations.

I think that this is to prevent adding a function in a resolve chain and affect as less as possible in the original behaviour.